File: app/src/main/java/com/rollapp/shared/ui/components/States.kt
package com.rollapp.shared.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.CloudOff
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.rollapp.shared.core.AppError

/**
 * Empty and error states share one shape: a mark, a plain sentence, and — when there
 * is something the user can actually do — exactly one button. An error that offers no
 * action gets no button rather than a dead "OK".
 */
@Composable
fun EmptyState(
    icon: ImageVector,
    title: String,
    body: String,
    modifier: Modifier = Modifier,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(horizontal = 40.dp),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            modifier = Modifier
                .size(72.dp)
                .clip(CircleShape)
                .background(MaterialTheme.colorScheme.surfaceVariant),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(32.dp)
            )
        }
        Spacer(Modifier.height(20.dp))
        Text(
            text = title,
            style = MaterialTheme.typography.titleLarge,
            textAlign = TextAlign.Center
        )
        Spacer(Modifier.height(8.dp))
        Text(
            text = body,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center
        )
        if (actionLabel != null && onAction != null) {
            Spacer(Modifier.height(24.dp))
            Button(onClick = onAction) { Text(actionLabel) }
        }
    }
}

@Composable
fun ErrorState(
    error: AppError,
    modifier: Modifier = Modifier,
    onRetry: (() -> Unit)? = null
) {
    EmptyState(
        icon = Icons.Rounded.CloudOff,
        title = error.message ?: "Something went wrong",
        body = errorGuidance(error),
        modifier = modifier,
        actionLabel = if (error.isRetryable && onRetry != null) "Try again" else null,
        onAction = onRetry.takeIf { error.isRetryable }
    )
}

/** The second line: what the user should understand, not a restatement of the error. */
private fun errorGuidance(error: AppError): String = when (error) {
    AppError.Offline -> "WeWere will catch up as soon as you're back on a network."
    AppError.Timeout -> "The connection is slow right now."
    AppError.PermissionDenied -> "Ask an admin to invite you again."
    AppError.GroupNotFound -> "It may have been deleted by an admin."
    AppError.RemovedFromGroup -> "An admin removed you from this group."
    AppError.InvalidInviteCode -> "Double-check the code — they're six characters."
    AppError.InviteExpired -> "Ask whoever invited you for a fresh link."
    AppError.StorageQuotaExceeded -> "This group has hit its storage limit."
    AppError.NotAuthenticated -> "Sign in to keep going."
    else -> "Give it another go in a moment."
}

/** A one-line, dismissible failure inside an otherwise working screen. */
@Composable
fun InlineError(
    message: String,
    modifier: Modifier = Modifier,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(MaterialTheme.shapes.medium)
            .background(MaterialTheme.colorScheme.errorContainer)
            .padding(start = 16.dp, end = 8.dp, top = 10.dp, bottom = 10.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = message,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onErrorContainer,
            modifier = Modifier.weight(1f)
        )
        if (actionLabel != null && onAction != null) {
            Spacer(Modifier.width(8.dp))
            TextButton(onClick = onAction) { Text(actionLabel) }
        }
    }
}

@Composable
fun LoadingState(modifier: Modifier = Modifier) {
    Box(modifier = modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        CircularProgressIndicator(strokeWidth = 2.5.dp)
    }
}

File: ios/WeWere/UI/Components/States.swift
import SwiftUI

/// Empty and error states share one shape: a mark, a plain sentence, and — when there
/// is something the user can actually do — exactly one button. An error that offers no
/// action gets no button rather than a dead "OK".
struct EmptyState: View {
    let icon: String
    let title: String
    let body_: String
    var actionLabel: String? = nil
    var onAction: (() -> Void)? = nil

    var body: some View {
        VStack(spacing: 0) {
            ZStack {
                Circle().fill(Raised)
                Image(systemName: icon)
                    .font(.system(size: 28, weight: .medium))
                    .foregroundStyle(IvoryMuted)
            }
            .frame(width: 72, height: 72)
            Spacer().frame(height: 20)
            Text(title).rollStyle(RollType.titleLarge, color: Ivory).multilineTextAlignment(.center)
            Spacer().frame(height: 8)
            Text(body_).rollStyle(RollType.bodyMedium, color: IvoryMuted).multilineTextAlignment(.center)
            if let actionLabel, let onAction {
                Spacer().frame(height: 24)
                FilledButton(text: actionLabel, action: onAction)
            }
        }
        .padding(.horizontal, 40)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

struct ErrorState: View {
    let error: AppError
    var onRetry: (() -> Void)? = nil

    var body: some View {
        EmptyState(
            icon: RollIcon.cloudOff,
            title: error.message ?? "Something went wrong",
            body_: errorGuidance(error),
            actionLabel: error.isRetryable && onRetry != nil ? "Try again" : nil,
            onAction: error.isRetryable ? onRetry : nil
        )
    }
}

/// The second line: what the user should understand, not a restatement of the error.
private func errorGuidance(_ error: AppError) -> String {
    switch error {
    case .offline: return "WeWere will catch up as soon as you're back on a network."
    case .timeout: return "The connection is slow right now."
    case .permissionDenied: return "Ask an admin to invite you again."
    case .groupNotFound: return "It may have been deleted by an admin."
    case .removedFromGroup: return "An admin removed you from this roll."
    case .invalidInviteCode: return "Double-check the code — they're six characters."
    case .inviteExpired: return "Ask whoever invited you for a fresh link."
    case .storageQuotaExceeded: return "This roll has hit its storage limit."
    case .notAuthenticated: return "Sign in to keep going."
    default: return "Give it another go in a moment."
    }
}

/// A one-line, dismissible failure inside an otherwise working screen.
struct InlineError: View {
    let message: String
    var actionLabel: String? = nil
    var onAction: (() -> Void)? = nil

    var body: some View {
        HStack(spacing: 8) {
            Text(message)
                .rollStyle(RollType.bodyMedium, color: OnErrorContainer)
                .frame(maxWidth: .infinity, alignment: .leading)
            if let actionLabel, let onAction {
                TextButton(text: actionLabel, color: OnErrorContainer, action: onAction)
            }
        }
        .padding(.leading, 16).padding(.trailing, 8).padding(.vertical, 10)
        .background(RoundedRectangle(cornerRadius: RollShapes.medium).fill(ErrorContainer))
    }
}

struct LoadingState: View {
    var body: some View {
        ProgressView().tint(Gold)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

/// M3 circular progress: an indeterminate spinner, or a determinate gold arc.
struct GoldRing: View {
    var progress: Double? = nil
    var size: CGFloat = 34
    var lineWidth: CGFloat = 3
    var track: Color = Ivory.opacity(0.2)

    var body: some View {
        if let progress {
            ZStack {
                Circle().stroke(track, lineWidth: lineWidth)
                Circle()
                    .trim(from: 0, to: min(max(progress, 0), 1))
                    .stroke(Gold, style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
                    .rotationEffect(.degrees(-90))
                    .animation(.easeOut(duration: 0.2), value: progress)
            }
            .frame(width: size, height: size)
        } else {
            ProgressView().tint(Gold).frame(width: size, height: size)
        }
    }
}

/// M3 linear progress: a thin gold bar on a raised track.
struct GoldBar: View {
    let fraction: Double

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(Raised)
                Capsule().fill(Gold).frame(width: geo.size.width * min(max(fraction, 0), 1))
            }
        }
        .frame(height: 4)
        .animation(.easeOut(duration: 0.2), value: fraction)
    }
}
